devel/style_guide.rst

```diff
@@ -0,0 +1,164 @@
+
+Coding conventions and style guide
+----------------------------------
+
+Basic Formatting and Indentation
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+The base formatting for all Zeek C++ code, including C++ code inside BIF files, follows the Whitesmiths coding style (https://en.wikipedia.org/wiki/Indentation_style#Whitesmiths_style).
+
+Tabs vs. Spaces
+^^^^^^^^^^^^^^^
+
+Use tabs for indentation and spaces for alignment. An example for alignment is below.
+
+.. sourcecode:: c++
+
+  |--\t--|void FunctionWithALongName(int argument1,
+  |--\t--|...........................int argument2);
+
+Tabs are used for the second line to line up with the start of the line above it, and then spaces are used to force the alignment with the line above it.
+
+File naming
+^^^^^^^^^^^
+
+Header files should always use the `.h` extension. Implementation files should always use the `.cc` extension.
+
+Include guards
+^^^^^^^^^^^^^^
+
+All headers should start with a `#pragma once` line to guard against duplicate includes. Avoid using `#ifndef`/`#define` include guards.
+
+Braces
+^^^^^^
+
+For multi-line blocks, braces should start on the line after the construct. Single-line blocks can remove the braces. For single-line function definitions the braces can be on the same line as the function definition, with a tab between the close of the function definition and the opening brace.
+
+Multi-line blocks
+*****************
+
+.. sourcecode:: c++
+
+  if ( true )
+	{
+	DoSomething();
+	DoSomethingElse();
+	}
+
+Single-line functions
+*********************
+
+.. sourcecode:: c++
+
+  bool Foo()	{ return false; }
+
+Note that for single-line functions there should be a tab between the closing `)` and the opening `{`.
+
+
+Function and variable naming
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Type names (classes, enums, structs, etc) and function names should always be `CamelCase`. Variable names, including member variables, should always be `snake_case`. Prefer using more descriptive variable names, except for counter variables.
+
+Including files
+^^^^^^^^^^^^^^^
+
+Include files in both headers and implementation files should be ordered as follows:
+
+- C includes such as `stdio.h`
+- C++ includes such as `string` and `vector`
+- Local include headers from Zeek
+
+Use angle braces around the file name for anything not coming directly from the Zeek code base. This includes any system headers and any external libraries. Use quotes around the file name for anything coming from the Zeek code base. This includes "external" libraries like Broker, since they are part of the Zeek code distribution.
```

Comment:
We currently use `<broker/data.hh>`, so I'd just make the rule be "use angle braces for anything that *can* be referring to a file outside the code distribution, even if typically it does refer to a file within the Zeek source tree because it's embedded for convenience (e.g. Broker/CAF)".